### PR TITLE
Feature mes 1401 view journal when not connected to network

### PR DESCRIPTION
**Pull request by rossfauldsbjss**
MES 1401 - Journal should be viewable from locally stored data when offline.

## Pull Request checklist:

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist:

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop

## Sign off process checklist:

- [x] Code has been tested manually
- [x] Tests are passing
- [x] PR link added to JIRA
- [x] Reviewers selected in Github

- [x] Tested by QA
- [x] PO's approval

### CODE REVIEW

File: src/providers/app-config/app-config.ts [hunk 1]
```diff
@@ -30,11 +36,27 @@ export class AppConfigProvider {
...
+      } else {
+        this.getOfflineConfig()
+        .then(data => resolve(data))
+        .catch(error => reject(error));
+      }
+    })
+
+  private getOfflineConfig = (): Promise<any> =>
+    new Promise((resolve, reject) => {
+      this.dataStore.getItem('CONFIG')
```
Comment (ammarhaiderbjss) [inline src/providers/app-config/app-config.ts]: Looks like this is already a promise so doesn't need to be wrapped in one too..

File: src/providers/journal/journal.ts [hunk 2]
```diff
@@ -5,28 +5,57 @@ import { UrlProvider } from '../url/url';
...
+    return this.getOfflineJournal();
+  }
+
+  getOfflineJournal(): Observable<ExaminerWorkSchedule> {
+    return from(this.getAndConvertOfflineJournal());
+  }
+
+  getAndConvertOfflineJournal(): Promise<ExaminerWorkSchedule> {
+    return new Promise<ExaminerWorkSchedule>((resolve, reject) => {
+      this.dataStore.getItem('JOURNAL').then((data) => {
```
Comment (ammarhaiderbjss) [inline src/providers/journal/journal.ts]: Again here, wrapping a promise in a promise

File: src/providers/app-config/app-config.ts (hunk 1 above)
Comment (johnmorleybjss) [inline src/providers/app-config/app-config.ts]: Could we rename `getOfflineConfig` to something like `getCachedRemoteConfig` - don't want us to get confused with config that's baked into the IPA.

File: src/providers/data-store/data-store.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+    this.secureContainer = container;
+  }
+
+  getSecureContainer(): SecureStorageObject {
+    return this.secureContainer;
+  }
+
+  getKeys(): any {
+    if (!this.secureContainer) {
+      return Promise.resolve('getKeys');
```
Comment (johnmorleybjss) [inline src/providers/data-store/data-store.ts]: Blank array would probably be better here and then you can put a type other than `any` on the method (keeps the type uniform).

File: src/providers/network-state/network-state.ts [hunk 4]
```diff
@@ -46,6 +43,9 @@ export class NetworkStateProvider {
   }
 
   public getNetworkState(): ConnectionStatus {
+    if (!this.networkStatus$) {
```
Comment (johnmorleybjss) [inline src/providers/network-state/network-state.ts]: Is this a workaround for running in browser? I don't understand the reasoning, so might be worth a comment explaining why.

File: src/providers/app-config/app-config.ts (hunk 1 above)
Comment (rossfauldsbjss) [reply src/providers/app-config/app-config.ts]: changed

File: src/providers/journal/journal.ts (hunk 2 above)
Comment (rossfauldsbjss) [reply src/providers/journal/journal.ts]: changed

File: src/pages/login/login.ts [hunk 5]
```diff
@@ -44,6 +42,7 @@ export class LoginPage extends BasePageComponent {
 
     // Trigger Authentication if this isn't a logout and is an ios device
     if (!this.hasUserLoggedOut && this.isIos()) {
+      console.log('trying to login');
```
[the remark this answers is not in the document]
Comment (rossfauldsbjss) [reply src/pages/login/login.ts]: removed

File: src/providers/app-config/app-config.ts (hunk 1 above)
Comment (rossfauldsbjss) [reply src/providers/app-config/app-config.ts]: done

File: src/providers/data-store/data-store.ts (hunk 3 above)
Comment (rossfauldsbjss) [reply src/providers/data-store/data-store.ts]: done

File: src/providers/network-state/network-state.ts (hunk 4 above)
Comment (rossfauldsbjss) [reply src/providers/network-state/network-state.ts]: It is - and I'm going to add some JSdocs in this file to explain.

Comment (rossfauldsbjss) [reply src/providers/network-state/network-state.ts]: Added jsdoc for this file.

**Comment by Cavalier07** [comment]
Was looking at the ticket and the work looks but , however I wanted to question the approach to local data storage.

In this implementation we store each piece of data (journal and config) and mange that individually. Would a better approach be to do this more at an app level and store the redux store in the database. and when we update the store an effect is triggered to push that to local storage. When we load the app if offline we only need to pull one thing from storage not 3 or 4 different things.

With some of the upcoming tickets like recovering from a crash , reducing loss of state/data , I can see this alternative approach making our lives easier and potentially causing less work (as test-report offline will 90% be done as well) 

I wanted to open up the conversation and see what your thoughts were

**Comment by ammarhaiderbjss** [comment]
> Was looking at the ticket and the work looks good , however I wanted to question the approach to local data storage.
> 
> In this implementation we store each piece of data (journal and config) and mange that individually. Would a better approach be to do this more at an app level and store the redux store in the database. and when we update the store an effect is triggered to push that to local storage. When we load the app if offline we only need to pull one thing from storage not 3 or 4 different things.
> 
> With some of the upcoming tickets like recovering from a crash , reducing loss of state/data , I can see this alternative approach making our lives easier and potentially causing less work (as test-report offline will 90% be done as well)
> 
> Work would be needed to put the config in the store but thats really where it should be anyway
> 
> I wanted to open up the conversation and see what your thoughts were

Agreed. This makes a lot of sense and something we should definitely do. However, I think it sits outside the scope of this work. The next time we save something to the device, we'll look at introducing this new generic pattern and then we can refactor the config and journal stuff to fall in line.

**Comment by szabolcs-keresztes** [comment]
We already have a bunch of things to refactor. My intuition says that we should start doing things the right way from the beginning as much as we can. Because the 6 sprints wall in the office will not allow us to do refactoring any time soon.

**Comment by szabolcs-keresztes** [comment]
Also would like to add that the unit testing framework what we have in place makes refactoring very challenging.

**Pull request merged**